featurebyte/service/feature_materialize.py

```diff
@@ -0,0 +1,332 @@
+    """
+
+    graph: QueryGraph
+    node_names: List[str]
+    output_column_names: List[str]
+
+
+@dataclass
+class OnlineEnabledFeaturesMetadata:
+    """
+    Information about online enabled features having the same primary entity that should be
+    published together
+    """
+
+    features: List[FeatureModel]
+    online_store_table_names: List[str]
+    primary_entities: List[EntityModel]
+    ingest_graph_and_node_names: Tuple[QueryGraph, List[str]] = field(init=False)
+    output_column_names: List[str] = field(init=False)
+
+    def __post_init__(self) -> None:
+        ingest_graph_metadata = self._get_combined_ingest_graph()
+        self.ingest_graph_and_node_names = (
+            ingest_graph_metadata.graph,
+            ingest_graph_metadata.node_names,
+        )
+        self.output_column_names = ingest_graph_metadata.output_column_names
+
+    @property
+    def serving_names(self) -> List[str]:
+        """
+        Returns serving names for the primary entities
+
+        Returns
+        -------
+        List[str]
+        """
+        return sorted([entity.serving_names[0] for entity in self.primary_entities])
+
+    @property
+    def primary_entity_ids(self) -> List[ObjectId]:
+        """
+        Returns primary entity ids
+
+        Returns
+        -------
+        List[ObjectId]
+        """
+        return sorted([entity.id for entity in self.primary_entities])
+
+    @property
+    def entity_universe_expr(self) -> Optional[Select]:
+        """
+        Returns a SQL expression that selects the distinct entity serving names across all online
+        store tables
+
+        Returns
+        -------
+        Optional[Select]
+        """
+        if not self.online_store_table_names:
+            return None
+
+        # select distinct serving names across all online store tables
+        distinct_serving_names_from_tables = [
+            select(*[quoted_identifier(serving_name) for serving_name in self.serving_names])
+            .distinct()
+            .from_(table_name)
+            for table_name in self.online_store_table_names
+        ]
+
+        union_expr = distinct_serving_names_from_tables[0]
+        for expr in distinct_serving_names_from_tables[1:]:
+            union_expr = expressions.Union(this=expr, distinct=True, expression=union_expr)  # type: ignore
+
+        return union_expr
+
+    def _get_combined_ingest_graph(self) -> OfflineIngestGraphMetadata:
+        """
+        Returns a combined ingest graph and related information for all features
+
+        Returns
+        -------
+        OfflineIngestGraphMetadata
+        """
+        local_query_graph = QueryGraph()
+        output_nodes = []
+        output_column_names = []
+
+        for feature in self.features:
+            offline_ingest_graphs = feature.extract_offline_store_ingest_query_graphs()
+            for offline_ingest_graph in offline_ingest_graphs:
```

Comment:
Does it make sense to you if we include `online_store_table_name` in offline ingest query graph? The rationale of doing this is to ease the feast feature view construction.

Reply:
I was thinking of keeping track of that information in a separate collection such as

```python
class OfflineStoreFeatureTable(FeatureByteBaseDocumentModel):

    name: str
    feature_ids: List[PydanticObjectId]
    primary_entity_ids: List[PydanticObjectId]
    # ... and other attributes
```

which will be updated on each feature's enabling / disabling. But I think including the online store table name in offline ingest query graph also makes sense. Which one do you prefer?

In either case, I think we will still have the above collection to simplify `FeatureMaterializeService` and the orchestration of scheduled tasks for feature publishing.